src/coreclr/jit/lclmasks.cpp

```diff
@@ -0,0 +1,480 @@
+        currentCost += cost;
+    }
+    else
+    {
+        // Switching would require adding a convert.
+        weight_t cost = isStore ? costOfConvertVectorToMask : costOfConvertMaskToVector;
+        cost *= blockWeight;
+
+        JITDUMP("Incrementing switchCost by %.2f. ", cost);
+        switchCost += cost;
+    }
+    DumpTotalWeight();
+}
+
+//-----------------------------------------------------------------------------
+// CacheSimdTypes: Cache the simd types of a hwintrinsic
+//
+// Arguments:
+//     op - The HW intrinsic to cache
+//
+void LclMasksWeight::CacheSimdTypes(GenTreeHWIntrinsic* op)
+{
+    CorInfoType newSimdBaseJitType = op->GetSimdBaseJitType();
+    unsigned    newSimdSize        = op->GetSimdSize();
+
+    assert((newSimdBaseJitType != CORINFO_TYPE_UNDEF));
+
+    simdBaseJitType = newSimdBaseJitType;
+    simdSize        = newSimdSize;
+}
+
+//-----------------------------------------------------------------------------
+// LclMasksCheckVisitor: Find all lcl var definitions and uses. For each one, update the weighting.
+//
+class LclMasksCheckVisitor final : public GenTreeVisitor<LclMasksCheckVisitor>
+{
+public:
+    enum
+    {
+        DoPostOrder       = true,
+        UseExecutionOrder = true
+    };
+
+    LclMasksCheckVisitor(Compiler* compiler, weight_t bbWeight, LclMasksWeightTable* weightsTable)
+        : GenTreeVisitor<LclMasksCheckVisitor>(compiler)
+        , bbWeight(bbWeight)
+        , weightsTable(weightsTable)
+    {
+    }
+
+    Compiler::fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTreeHWIntrinsic* convertOp = nullptr;
+
+        bool isLocalStore  = false;
+        bool isLocalUse    = false;
+        bool hasConversion = false;
+
+        switch ((*use)->OperGet())
+        {
+            case GT_STORE_LCL_VAR:
+                isLocalStore = true;
+
+                // Look for:
+                //      use:STORE_LCL_VAR(ConvertMaskToVector(x))
+
+                if ((*use)->AsLclVar()->Data()->OperIsConvertMaskToVector())
+                {
+                    convertOp     = (*use)->AsLclVar()->Data()->AsHWIntrinsic();
+                    hasConversion = true;
+                }
+                break;
+
+            case GT_LCL_VAR:
+                isLocalUse = true;
+
+                // Look for:
+                //      user:ConvertVectorToMask(use:LCL_VAR(x)))
+
+                if (user->OperIsConvertVectorToMask())
+                {
+                    convertOp     = user->AsHWIntrinsic();
+                    hasConversion = true;
+                }
+                break;
+
+            default:
+                break;
+        }
+
+        if (isLocalStore || isLocalUse)
+        {
+            GenTreeLclVarCommon* lclOp = (*use)->AsLclVarCommon();
+
+            // Get the existing weighting (if any).
+            LclMasksWeight weight;
+            weightsTable->Lookup(lclOp->GetLclNum(), &weight);
```

Comment:
This can use `LookupPointerOrAdd` instead to avoid multiple unnecessary hash table lookups and `LclMasksWeight` copies.

Reply:
> This can use LookupPointerOrAdd instead to avoid multiple unnecessary hash table lookups and LclMasksWeight copies.

That doesn't seem to work:

```
            LclMasksWeight weight;
            weightsTable->LookupPointerOrAdd(lclOp->GetLclNum(), weight);
```

I think it ends up storing a pointer to within the local stack. Resulting in that every entry then ends up pointing to the same stack place.

In this dump, it sets every local, then in the bottom section, it reads the weighting, and gets the same value for every local:

```
Local var V06 at [000047] has no conversion. Incrementing switchCost by 20000.00. Weighting: {0.00c 20000.00s}
Local store V00 at [000010] has no conversion. Incrementing switchCost by 40000.00. Weighting: {0.00c 40000.00s}
Local var V11 at [000103] has no conversion. Incrementing switchCost by 20000.00. Weighting: {0.00c 20000.00s}
Local store V01 at [000014] has no conversion. Incrementing switchCost by 40000.00. Weighting: {0.00c 40000.00s}
Local store V05 at [000015] has mask conversion. Incrementing currentCost by 20000.00. Weighting: {20000.00c 0.00s}
Local var V05 at [000017] has mask conversion. Incrementing currentCost by 40000.00. Weighting: {40000.00c 0.00s}
Local var V00 at [000018] has no conversion. Incrementing switchCost by 20000.00. Weighting: {0.00c 20000.00s}
Local var V01 at [000019] has no conversion. Incrementing switchCost by 20000.00. Weighting: {0.00c 20000.00s}
Local store V02 at [000023] has no conversion. Incrementing switchCost by 40000.00. Weighting: {0.00c 40000.00s}
Local var V05 at [000016] has no conversion. Incrementing switchCost by 20000.00. Weighting: {0.00c 20000.00s}
Local store V03 at [000025] has no conversion. Incrementing switchCost by 40000.00. Weighting: {0.00c 40000.00s}
Local var V02 at [000026] has no conversion. Incrementing switchCost by 20000.00. Weighting: {0.00c 20000.00s}
Local var V03 at [000027] has no conversion. Incrementing switchCost by 20000.00. Weighting: {0.00c 20000.00s}

Local var V06 at [000047] will not be converted. Weighting: {0.00c 20000.00s}
Local store V00 at [000010] will not be converted. Weighting: {0.00c 20000.00s}
Local var V11 at [000103] will not be converted. Weighting: {0.00c 20000.00s}
Local store V01 at [000014] will not be converted. Weighting: {0.00c 20000.00s}
Local store V05 at [000015] will not be converted. Weighting: {0.00c 20000.00s}
Local var V05 at [000017] will not be converted. Weighting: {0.00c 20000.00s}
Local var V00 at [000018] will not be converted. Weighting: {0.00c 20000.00s}
Local var V01 at [000019] will not be converted. Weighting: {0.00c 20000.00s}
Local store V02 at [000023] will not be converted. Weighting: {0.00c 20000.00s}
Local var V05 at [000016] will not be converted. Weighting: {0.00c 20000.00s}
Local store V03 at [000025] will not be converted. Weighting: {0.00c 20000.00s}
Local var V02 at [000026] will not be converted. Weighting: {0.00c 20000.00s}
Local var V03 at [000027] will not be converted. Weighting: {0.00c 20000.00s}
```

Reply:
> returns a pointer

That fixed it.